File: internal/core/application/usecases/generate_payment.go
package usecases

import (
	"context"
	"time"

	"github.com/fiap/ms-payment/internal/adapters/driver/dto"
	"github.com/fiap/ms-payment/internal/core/domain/entities"
	"github.com/fiap/ms-payment/internal/core/domain/ports"
	"github.com/fiap/ms-payment/pkg/utils"
)

type GeneratePaymentUseCase struct {
	paymentRepo      ports.PaymentRepository
	generateTokenMP  ports.MercadoPagoGenerateToken
	generateQRCodeMP ports.MercadoPagoGenerateQRCode
	paymentTopic     ports.PaymentTopic
}

func NewGeneratePaymentUseCase(
	paymentRepo ports.PaymentRepository,
	generateTokenMP ports.MercadoPagoGenerateToken,
	generateQRCodeMP ports.MercadoPagoGenerateQRCode,
	paymentTopic ports.PaymentTopic,
) *GeneratePaymentUseCase {
	return &GeneratePaymentUseCase{
		paymentRepo:      paymentRepo,
		generateTokenMP:  generateTokenMP,
		generateQRCodeMP: generateQRCodeMP,
		paymentTopic:     paymentTopic,
	}
}

func (u *GeneratePaymentUseCase) Execute(ctx context.Context, input dto.GeneratePaymentInputDTO) error {
	if input.ExternalId == "" {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "missing external ID")
		return nil
	}

	if input.Amount <= 0 {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "missing amount")
		return nil
	}

	if input.Description == "" {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "missing description")
		return nil
	}

	payment, err := u.paymentRepo.GetByExternalId(ctx, input.ExternalId)
	if err != nil {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "failed to get payment by external ID")
		return utils.HTTPInternalServerError("failed to get payment by external ID")
	}

	if payment.Id != "" {
		if payment.Status == entities.PaymentStatus_Paid {
			_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "payment already paid")
			return utils.HTTPConflict("payment already paid")
		}

		if time.Unix(payment.ExpiresAt, 10).After(time.Now()) {
			_ = u.paymentTopic.PublishPaymentCreationSuccess(ctx, input.ExternalId, payment.Code)
			return nil
		}

		err := u.paymentRepo.Delete(ctx, payment.ExternalId)
		if err != nil {
			_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "failed to delete expired payment")
			return utils.HTTPInternalServerError("failed to delete expired payment")
		}
	}

	token, err := u.generateTokenMP.GenerateToken(ctx)
	if err != nil {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "failed to generate MercadoPago token")
		return utils.HTTPPreconditionFailed("failed to generate MercadoPago token")
	}

	qrCode, err := u.generateQRCodeMP.GenerateQRCode(ctx, input.ExternalId, input.Amount, input.Description, token)
	if err != nil {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "failed to generate MercadoPago QR Code")
		return utils.HTTPPreconditionFailed("failed to generate MercadoPago QR Code")
	}

	newPayment := entities.NewPayment(input.ExternalId, input.Amount, input.Description, qrCode)

	err = u.paymentRepo.Create(ctx, *newPayment)
	if err != nil {
		_ = u.paymentTopic.PublishPaymentCreationFailure(ctx, input.ExternalId, "failed to create new payment")
		return utils.HTTPInternalServerError("failed to create new payment")
	}

	_ = u.paymentTopic.PublishPaymentCreationSuccess(ctx, input.ExternalId, newPayment.Code)

	return nil
}

File: src/core/application/usecases/generate_payment.ts
import { IMercadoPagoGenerateQRCode } from "@ports/generate_qrcode_mp"
import { IMercadoPagoGenerateToken } from "@ports/generate_token_mp"
import { GeneratePaymentInputDTO } from "@dto/generate_payment"
import { IPaymentRepository } from "@ports/payment_repository"
import { PaymentStatus } from "@entities/payment.types"
import { IPaymentTopic } from "@ports/payment_topic"
import { Payment } from "@entities/payment"

export class GeneratePaymentUseCase {
	private paymentRepository: IPaymentRepository
	private generateTokenMP: IMercadoPagoGenerateToken
	private generateQRCodeMP: IMercadoPagoGenerateQRCode
	private paymentTopic: IPaymentTopic

	constructor(
		paymentRepository: IPaymentRepository,
		generateTokenMP: IMercadoPagoGenerateToken,
		generateQRCodeMP: IMercadoPagoGenerateQRCode,
		paymentTopic: IPaymentTopic,
	) {
		this.paymentRepository = paymentRepository
		this.generateTokenMP = generateTokenMP
		this.generateQRCodeMP = generateQRCodeMP
		this.paymentTopic = paymentTopic
	}

	async execute(params: GeneratePaymentInputDTO): Promise<void> {
		if (!params.external_id) {
			await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "missing external ID")
			return
		}

		if (params.amount <= 0) {
			await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "missing amount")
			return
		}

		if (!params.description) {
			await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "missing description")
			return
		}

		const existingPayment = await this.paymentRepository.getByExternalId(params.external_id)
		if (existingPayment) {
			if (existingPayment.status === PaymentStatus.Paid) {
				await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "payment already paid")
				return
			}

			if (existingPayment.expires_at > Date.now()) {
				await this.paymentTopic.publishPaymentCreationSuccess(params.external_id, existingPayment.code)
				return
			}

			await this.paymentRepository.delete(params.external_id)
		}

		const token = await this.generateTokenMP.generateToken()
		if (!token) {
			await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "failed to generate MercadoPago token")
			return
		}

		const qrCode = await this.generateQRCodeMP.generateQRCode(
			params.external_id,
			params.amount,
			params.description,
			token,
		)
		if (!qrCode) {
			await this.paymentTopic.publishPaymentCreationFailure(params.external_id, "failed to generate MercadoPago QR Code")
			return
		}

		const newPayment = Payment.create(
			params.external_id,
			params.amount,
			params.description,
			qrCode,
		)

		await this.paymentRepository.create(newPayment)

		await this.paymentTopic.publishPaymentCreationSuccess(params.external_id, newPayment.code)
	}
}